### ISSUE TITLE
Avoid "str | bytes" where names or text are meant

### ISSUE THREAD

**Issue by matangover** [lxml/lxml-stubs#29]
Right now `_Element.text` is typed as `Optional[Union[str, bytes]]`. However according to the [lxml FAQ]:(https://lxml.de/3.6/FAQ.html):
> In Python 3, lxml always returns Unicode strings for text and names, as does ElementTree.

Having to deal with `bytes` makes all code accessing `_Element.text` unnecessarily cumbersome.

**Comment by matangover** [comment]
This is also relevant to the attrib dictionary. I'm not sure whether it can actually contain `bytes` values (couldn't find info in docs). However right now there is a discrepancy - because `get` always returns a `str`. So one of those is wrong.

**Comment by scoder** [comment]
The thing is, as long as users are still running their code in Py2, and writing code that works with Py2 and Py3, both need to be supported. Users who only target Py2 and not Py3 at all would probably also not use static type checkers, but legacy Py2/3 code should be supported.

That said, yes, the current declarations are inconsistent.

And the fact that legacy code needs to be supported doesn't mean that `str` wouldn't be sufficient. lxml does not use `bytes` as in "sequence of bytes" but only as byte encoded textual data, i.e. Py2 `str`, not Py3 `bytes`.

I think it's worth tying a switch to plain `str` and seeing if that really breaks anything.

PR welcome.

**Comment by matangover** [comment]
> I think it's worth trying a switch to plain str and seeing if that really breaks anything.

It would break correctness on Python 2 because on Python 2 these attributes may return either str or unicode. Can we drop Python 2 compatibility of the typings? If yes, we can switch to `str`.

If you want to keep Python 2 compatibility, we can use `Union[typing.Text, str]` or use a [version-dependent](https://mypy.readthedocs.io/en/stable/common_issues.html#python-version-and-system-platform-checks) type alias.

**Comment by abelcheung** [comment]
@matangover Agreed it's a good idea to drop py2 support for annotations. Personally I don't know anybody who's writing new code on py2 now, lest using annotation on py2 project.

However, the removal of `bytes` is mostly for method / function return type and object attributes. `bytes` as input argument is still in use, I'm not sure dropping that is good idea or not.

**Comment by matangover** [comment]
Agreed, `str` should be used for places where "names or text" are returned / accepted.

**Comment by macro1** [comment]
Per https://github.com/python/typing/issues/208 an instance of Python 2 `str` is a subtype of `unicode`. Are we over thinking this, and would `typing.Text` just work for 2.7 type checking? It seems to me the only thing we need to do is indicate for Python 3 where `bytes` are acceptable as an input as well.

**Comment by 5j9** [comment]
The latest release of `lxml` only supports 3.6+. [Support for Python 2.7 and Python versions < 3.6 was removed](https://lxml.de/5.3/changes-5.3.0.html) in v5.1.0 (2024-01-05).
